### PR TITLE
File reader script

### PR DESCRIPTION
**Pull request by aeb0**
Intended for generating csvs that can be easily read into postgres.
After looking at the slack export, it looked like there are three logical pieces of data that lend themselves to SQL tables.  Thus the script ouputs one csv file per table.

TODO:
1) Remove duplicates from the user before writing to file
2) Escape commas in messages and attachments since they will screw up the comma delimiting of the csv.  Or delimit with tab or something

### CODE REVIEW

File: scripts/json-to-csv.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  const outputPathSuffix = cleanedPath.concat('.csv');
+
+  console.log(`writing messages to csv: ${__dirname.concat('/messages-'.concat(outputPathSuffix))}`);
+  fs.writeFileSync(__dirname.concat('/messages-'.concat(outputPathSuffix)), messagesCSV, 'utf8');
+
+  console.log(`writing users to csv: ${__dirname.concat('/users-'.concat(outputPathSuffix))}`);
+  fs.writeFileSync(__dirname.concat('/users-'.concat(outputPathSuffix)), messagesCSV, 'utf8');
+
+  console.log(`writing attachments to csv: ${__dirname.concat('/attachments-'.concat(outputPathSuffix))}`);
+  fs.writeFileSync(__dirname.concat('/attachments-'.concat(outputPathSuffix)), messagesCSV, 'utf8');
```
Comment (mLuby) [inline scripts/json-to-csv.js]: No need for these `fs.writeFile`s to be sync as they don't have any after-effects.

Also do you mean to be saving the same messagesCSV to three places?

File: scripts/json-to-csv.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   *    Users
+   *    |usercode|username|
+   *
+   *
+   *    Attachments
+   *    |servicename|title|titlelink|text|fromurl|
+   */
+
+  parsedMessages.forEach((message) => {
```
Comment (mLuby) [inline scripts/json-to-csv.js]: This is a `reduce` operation, taking `message` objects and returning `messagesCSV`, `usersCSV`, and `attachmentsCSV` strings. 🤓

File: scripts/json-to-csv.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+
+
+function cleanSlackJSON(path, channel) {
+  if (arguments.length !== 2) {
+    console.error("Usage: cleanSlackJSON(file, channel)");
+    return;
+  }
+
+  const messages = fs.readFileSync(path, 'utf8');
```
Comment (mLuby) [inline scripts/json-to-csv.js:21]: Everything else could happen as the `async` version's callback, right?

File: scripts/json-to-csv.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
+ /*
+  *  One JSON file from Slack export in, 3 csv files out
+  *  We ignore parts of the data people are unlikely to care about in order
+  *  to save space: channel join messages, reactions, edit messages,
+  *  pinned items, certain attachment metadata
```
Comment (mLuby) [inline scripts/json-to-csv.js:5]: I for one feel reactions are important, but we can always add that in after MVP right?

File: sample-data/2016-11-03.json [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        "ts": "1478217394.000202"
+    },
+    {
+        "user": "U2YNU36TG",
+        "text": "<@U2YNU36TG|pmsandhu> has joined the channel",
+        "type": "message",
+        "subtype": "channel_join",
+        "ts": "1478218946.000203"
+    }
+]
```
Comment (mLuby) [inline sample-data/2016-11-03.json:53]: missing newline

File: scripts/json-to-csv.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   *
+   *    Attachments
+   *    |servicename|title|titlelink|text|fromurl|
+   */
+
+  parsedMessages.forEach((message) => {
+    if (message.subtype === 'channel_join' || message.subtype === 'pinned_item') return;
+
+    let messagesRow    = [];
```
Comment (dting) [inline scripts/json-to-csv.js]: `const`s here?

File: scripts/json-to-csv.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    if (message.user) {
+      messagesRow[0]                            = message.user.slice(0,10);
+      usersRow[0]                               = message.user.slice(0,10);
+      if (message.user.length > 10) usersRow[1] = message.user.slice(11);
+    }
+    if (message.text)      messagesRow[1] = message.text;
+    if (message.ts)        messagesRow[2] = message.ts;
+    if (message.purpose)   messagesRow[3] = true;
+    if (message.attachments) {
+      message.attachments.forEach((attachment) => {
```
Comment (dting) [inline scripts/json-to-csv.js]: Can there be multiple attachments per message? If so, does this just leave you with the last attachment's info in attachmentsRow?

File: scripts/json-to-csv.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+   *    |user|text|timestamp|purpose|channel|
+   *
+   *
+   *    Users
+   *    |usercode|username|
+   *
+   *
+   *    Attachments
+   *    |servicename|title|titlelink|text|fromurl|
+   */
```
Comment (dting) [inline scripts/json-to-csv.js:36]: Does this leave us with 3 separate `csv`s with no way to associate which message each attachment came from and duplicate entries in the users `csv`?

File: scripts/json-to-csv.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      messagesCSV += messagesRow.join(',').concat('\n');
+    }
+    if (usersRow.length) {
+      usersCSV += usersRow.join(',').concat('\n');
+    }
+    if (attachmentsRow.length) {
+      attachmentsRow[5] = channel;
+      attachmentsCSV += attachmentsRow.join(',').concat('\n');
+    }
+  });
```
Comment (dting) [inline scripts/json-to-csv.js]: I feel like we should be writing these out as they are processed or the entire contents of the file needs to be read into memory before we write anything out. I haven't done any real research but a library like https://github.com/maxogden/csv-write-stream would probably work.

File: scripts/json-to-csv.js (hunk 1 above)
Comment (aeb0) [reply scripts/json-to-csv.js]: Oops, no supposed to be messagesCSV, attachmentsCSV, usersCSV. fixed.

Comment (aeb0) [reply scripts/json-to-csv.js]: Regarding the sync vs async...My thinking was sync might be safer than async, as this script may be triggered inside of a loop that iterates over the thousands of json files in the slack export.  If its sync node will block on each iteration preventing JSON.parse from being called on a ton of shit at the same time and maybe causing memory issues.  

Probably not a serious concern though. Do you have a reason for suggesting async?

File: scripts/json-to-csv.js (hunk 4 above)
Comment (aeb0) [reply scripts/json-to-csv.js:5]: We could, also not a big deal to add them in right now.

File: scripts/json-to-csv.js (hunk 1 above)
Comment (mLuby) [reply scripts/json-to-csv.js]: That sounds like a good reason for sync.

**Comment by mLuby** [comment]
re removing dupes—can be done in bash pretty quickly.

``` bash
sort users-file.sql | uniq
```

A little trickier in JS.

``` js
dedup(usersCSV.split('\n')))
function dedup (listOfStrings) { return Object.keys(listOfStrings.reduce((uniques, string) => Object.assign(uniques, {[string]: true}), {})) }
```

re delimiting, also escape newlines?

File: scripts/json-to-csv.js (hunk 8 above)
Comment (dting) [reply scripts/json-to-csv.js:36]: I totally should have read the comments... Sorry I was in a rush this morning. Ignore the dupes remark.

**Comment by dting** [comment]
Hmm. I'm kind of wondering why even go to the trouble of writing this out to csv when we can just batch insert into pg. Are we going to use these files somewhere else? If not just inserting them directly into pg seems to make more sense to retain their relationships.

File: scripts/json-to-csv.js (hunk 2 above)
Comment (aeb0) [reply scripts/json-to-csv.js]: good point. fixed in refactor

File: scripts/json-to-csv.js (hunk 7 above)
Comment (aeb0) [reply scripts/json-to-csv.js]: there could be good, point and also yes that was a bug. fixed in refactor :)

**Comment by aeb0** [comment]
@dting can you explain more on if batch insert would work with json? And into a document store or relational?  (I am operating on the assumption we are going relational based on a couple of articles)  Googled around a bit but didn't find anything that helpful re: batch insert.

File: scripts/json-to-csv.js [hunk 10]
```diff
@@ -38,58 +32,89 @@ function cleanSlackJSON(path, channel) {
...
-    }
-    if (usersRow.length) {
-      usersCSV += usersRow.join(',').concat('\n');
-    }
-    if (attachmentsRow.length) {
-      attachmentsRow[5] = channel;
-      attachmentsCSV += attachmentsRow.join(',').concat('\n');
-    }
-  });
+   let messagesCSV = parsedMessages.reduce((accum, currMessage) => {
```
Comment (mLuby) [inline scripts/json-to-csv.js:38]: Close to what I meant. I think you can avoid some duplication via destructuring:

``` js
const {messagesCSV, usersCSV, attachmentsCSV} = parsedMessages.reduce((accum, currMessage) => {
```

**Review by mLuby — APPROVED** [review]

**Comment by aeb0** [comment]
Holding on merging this while we figure out if we can use [another solution](https://github.com/bsvingen/slack2pg) to handle importing more easily

**Comment by mLuby** [comment]
Any reason not to merge this and if that other solution pans out submit a new PR replacing this with that? Iterative! Agile!

**Pull request merged**